src/DS/metrics_extraction.py

```diff
@@ -0,0 +1,300 @@
+            content = file.read()
+            tree = ast.parse(content, filename=file_path)
+            logger.debug(f"Successfully parsed file: {file_path}")
+            return tree, content
+    except SyntaxError as e:
+        logger.error(f"Syntax error in file {file_path}: {e}")
+    except UnicodeDecodeError as e:
+        logger.error(f"Unicode decode error in file {file_path}: {e}")
+    except FileNotFoundError as e:
+        logger.error(f"File not found: {file_path}: {e}")
+    except Exception as e:
+        logger.error(f"Unexpected error processing file {file_path}: {e}")
+
+    return None
+
+
+def analyse_test_file(file_path):
+    """
+    Analyses a Python test file to extract the number of test cases,
+    assertions, presence of setup and teardown methods, and the complexity
+    of the test file.
+
+    Args:
+        file_path (str): The path to the test file.
+
+    Returns:
+           dict: A dictionary containing the analysis results.
+    """
+    logger.info(f"Starting analysis of test file: {file_path}")
+
+    result = {
+        "num_test_cases": 0,
+        "num_assertions": 0,
+        "has_setup": False,
+        "has_teardown": False,
+        "complexity": 0,
+    }
+
+    parsed_file = read_and_parse_file(file_path)
+    if parsed_file is None:
+        logger.warning(f"Analysis skipped for file: {file_path}")
+        return result
+
+    tree, _ = parsed_file
+
+    num_test_cases = 0
+    num_assertions = 0
+    has_setup = False
+    has_teardown = False
+    complexity = 0
+
+    for node in ast.walk(tree):  # Walks through all nodes in the AST
+        if isinstance(node, ast.FunctionDef):
+            if node.name.startswith("test_"):
+                num_test_cases += 1
+                complexity += 1  # Basic complexity, 1 per function
+                logger.debug(f"Found test case: {node.name}")
+
+                for body_node in node.body:  # Iterates over the body of the
+                    # function
+                    if isinstance(body_node, ast.If):
+                        complexity += 1  # Each 'if' branch increases complexity
+                        logger.debug(
+                            f"Incrementing complexity for 'if' in" f" {node.name}"
+                        )
+
+                    if isinstance(body_node, ast.Expr) and isinstance(
+                        body_node.value, ast.Call
+                    ):  # Checks if the node is an expression and a function
+                        # call.
+                        if (
+                            isinstance(body_node.value.func, ast.Name)
+                            and body_node.value.func.id == "assert"
+                        ):  # Checks if the function call is an assertion.
+                            num_assertions += 1
+                            logger.debug(f"Found assertion in {node.name}")
+
+            if node.name in ("setUp", "tearDown"):
```

Comment:
I'd recommend separating the processing and reporting of `setUp` and `tearDown` methods. The cost of doing so is small in terms of code complexity, the value of doing so is more than the cost of doing so in my view, otherwise line 173 loses information and there may be two of these identical messages for a single node (file) in the log which look more like duplicates than being unambiguous debug information.

Reply:
Thank you. This is done.